File: LeetCode/!0078. Subsets.cpp
/*
0078. Subsets

给定一组不含重复元素的整数数组。返回所有可能的子集。注意：结果不能包含重复子集。

思路：遍历该整数数组。将最终返回的结果集合记做result。
每次遍历到一个新数，result中就要新增加：由result中所有子集加上当前这个数构成的一系列新集合。
例如，输入[1,2,3]，遍历到了3，此时result里为[[], 1, 2, 12]。那么result里就要新增以下子集：1, 12, 23, 123。
*/

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> emptySet;
        result.push_back(emptySet);
        for (int i = 0; i < nums.size(); i++) {
            // 向当前结果集合里所有子集（除了空集）添加当前元素, 获得若干新子集，再将他们加入result中
            int currentResultSize = result.size();
            for (int j = 0; j < currentResultSize; j++) {
                vector<int> newSubset = result[j];
                newSubset.push_back(nums[i]);
                result.push_back(newSubset);
            }
        }
        return result;
    }
};

File: LeetCode/!0078. Subsets.py
'''
0078. Subsets

给定一组不含重复元素的整数数组 nums，返回该数组所有可能的子集（幂集）。

思路：实际上是回溯法求组合的变体。区别在于，这道题每次进入递归都要向res里添加当前结果，
而不仅仅是在全部取完候选数字后再存入res。
'''
from typing import List


class Solution:
    # cpp可以写在这里，但是py不行，力扣跑下一个测试不会清空这边上一次的记录
    # res = [[]]
    # chosen = []

    def backtrack(self, nums, start_index, res, chosen):
        res.append(chosen.copy())
        if (start_index == len(nums)):
            return
        # dfs遍历
        for i in range(start_index, len(nums)):
            chosen.append(nums[i])
            # dfs
            self.backtrack(nums, i + 1, res, chosen)
            # 回溯
            chosen.pop()


    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        chosen = []
        self.backtrack(nums, 0, res, chosen)
        return res
